lucene/core/src/java/org/apache/lucene/search/TopScoreDocCollector.java

```diff
@@ -139,7 +140,8 @@ public void collect(int doc) throws IOException {
 
           if (score > after.score || (score == after.score && doc <= afterDoc)) {
             // hit was collected on a previous page
-            if (totalHitsRelation == TotalHits.Relation.EQUAL_TO && hitsThresholdChecker.isThresholdReached()) {
+            if ((totalHitsRelation == TotalHits.Relation.EQUAL_TO || shouldUpdateMinScore())
```

Comment:
I am not really sure if we need `shouldUpdateMinScore` here -- it would be called anyway if the global threshold is reached. Is the objective to reduce the number of redundant calls to `updateMinCompetitiveScore` i.e. only call when we know that we have a better global score?

Reply:
The idea here is that currently we call updateMinCompetitiveScore only once when the threshold is reached, we also update the minimum score when the queue is full and a new competitive hit is inserted. However now that the minimum score can be updated by other collectors it is possible that the global minimum score changed while the local one remains unchanged (because we see only non-competitive hits). This change ensures that we update the current min score to take into account the new global minimum score if it is greater than the local one. I think it's important because otherwise we'll miss a lot of updates to the global minimum score (e.g. if the collector only see non-competitive hits).

Reply:
I am not sure if I understand. If a collector is only seeing non competitive hits, it would not be able to contribute to the global minimum score?

On the other hand, if the local minimum score is enough to filter out non competitive hits but not good enough to beat the current global minimum, we could run with the local score and let it be the sentinel of the collector's bottom score.

Reply:
Currently if the local queue if full and the collector only sees non-competitive hits we don't update the minimum competitive score. This is normal since the local minimum competitive score cannot changed on non-competitive hits, however it is possible that the global minimum score changed in the meantime and that the collector is missing a big update that propagated to the local scorer could make the skipping of hits more efficient. That's what the test here is doing, checking that we can update the minimum score from the global one.
I refactored this part a bit in my last commit but the idea is the same.

Reply:
Ok, I think you reverted this change in the latest iteration?

Reply:
Yep sorry, after running some benchmarks it seems that checking the global score on every non-competitive hits hurts the performance of constant score queries.
The gains are already significative without it so I restored the previous version.